### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by bohdan-mykhailenko**

### CODE REVIEW

File: src/figuresClasses.ts [hunk 1]
```diff
@@ -1,19 +1,94 @@
...
-}
+  public sides: number[] = [];
+
+  constructor(public color: Color, ...sides: number[]) {
+    const longestSide = Math.max(...sides);
+    const sumOfAllSides = sides.reduce((perimiter, side) => perimiter + side);
+    const sumOfSmallSides = sumOfAllSides - longestSide;
 
-export class Circle {
+    if (longestSide >= sumOfSmallSides
```
Comment (naziamoff) [inline src/figuresClasses.ts:22]:
```suggestion
    if (
      longestSide >= sumOfSmallSides
```

File: src/figuresClasses.ts [hunk 2]
```diff
@@ -1,19 +1,94 @@
...
+      || sides[0] <= 0 || sides[1] <= 0 || sides[2] <= 0
+    ) {
+      throw new Error('Triangle with given sides isn\'t exist');
+    }
```
Comment (naziamoff) [inline src/figuresClasses.ts:26]: let's divide this logic into two separate errors

File: src/figuresClasses.ts [hunk 3]
```diff
@@ -1,19 +1,94 @@
...
 }
 
-export class Triangle {
+export class Triangle implements Figure {
+  public shape: Shape = 'triangle';
 
-}
+  public sides: number[] = [];
+
+  constructor(public color: Color, ...sides: number[]) {
```
Comment (naziamoff) [inline src/figuresClasses.ts:17]: let's move the logic with throwing errors to a separate method - and call it inside the constructor

File: src/figuresClasses.ts [hunk 4]
```diff
@@ -1,19 +1,94 @@
...
+
+    return Math.floor(area * 100) / 100;
+  }
 }
 
-export class Rectangle {
+export class Circle implements Figure {
+  public shape: Shape = 'circle';
+
+  constructor(public color: Color, public radius: number) {
```
Comment (naziamoff) [inline src/figuresClasses.ts:52]:
```suggestion
  constructor(
    public color: Color, 
    public radius: number
  ) {
```

File: src/figuresClasses.ts [hunk 5]
```diff
@@ -1,19 +1,94 @@
...
+  constructor(public color: Color, ...sides: number[]) {
+    const longestSide = Math.max(...sides);
+    const sumOfAllSides = sides.reduce((perimiter, side) => perimiter + side);
+    const sumOfSmallSides = sumOfAllSides - longestSide;
 
-export class Circle {
+    if (longestSide >= sumOfSmallSides
+      || sides[0] <= 0 || sides[1] <= 0 || sides[2] <= 0
+    ) {
+      throw new Error('Triangle with given sides isn\'t exist');
```
Comment (naziamoff) [inline src/figuresClasses.ts:25]:
```suggestion
      throw new Error('Triangle with given sides doesn\'t exist');
```

**Review by naziamoff — CHANGES_REQUESTED** [review]

File: src/figuresClasses.ts [hunk 6]
```diff
@@ -1,19 +1,105 @@
...
+export class Triangle implements Figure {
+  public shape: Shape = 'triangle';
 
-}
+  public sides: number[] = [];
+
+  constructor(
+    public color: Color,
+    ...sides: number[]
+  ) {
```
Comment (33Orange) [inline src/figuresClasses.ts:20]: Not the best place to use the rest operator

```suggestion
  constructor(
    public color: Color,
    public a: number,
    public b: number,
    public c: number
  ) {
```

File: src/figuresClasses.ts [hunk 7]
```diff
@@ -1,19 +1,105 @@
...
+    const longestSide = Math.max(...sides);
+    const sumOfAllSides = sides.reduce((perimiter, side) => perimiter + side);
+    const sumOfSmallSides = sumOfAllSides - longestSide;
+
+    if (longestSide >= sumOfSmallSides) {
+      throw new Error('Triangle with given sides doesn\'t exist');
+    }
 
-export class Circle {
+    if (sides[0] <= 0 || sides[1] <= 0 || sides[2] <= 0) {
```
Comment (33Orange) [inline src/figuresClasses.ts:29]: That's why is not the best case to use rest here

File: src/figuresClasses.ts [hunk 8]
```diff
@@ -1,19 +1,105 @@
...
+    }
 
+    this.color = color;
+    this.sides = sides;
+  }
+
+  getArea(): number {
+    const sideA = this.sides[0];
+    const sideB = this.sides[1];
+    const sideC = this.sides[2];
```
Comment (33Orange) [inline src/figuresClasses.ts:40]: Same here

File: src/figuresClasses.ts [hunk 9]
```diff
@@ -1,19 +1,105 @@
...
-export function getInfo(figure) {
+export class Rectangle implements Figure {
+  public shape: Shape = 'rectangle';
+
+  public sides: number[] = [];
+
+  constructor(
+    public color: Color,
+    ...sides: number[]
+  ) {
```
Comment (33Orange) [inline src/figuresClasses.ts:85]:
```suggestion
  constructor(
    public color: Color,
    public width: number,
    public height: number
  ) {
```

File: src/figuresClasses.ts [hunk 10]
```diff
@@ -1,19 +1,105 @@
...
+
+    return Math.floor(area * 100) / 100;
+  }
 }
 
-export function getInfo(figure) {
+export class Rectangle implements Figure {
+  public shape: Shape = 'rectangle';
+
+  public sides: number[] = [];
```
Comment (33Orange) [inline src/figuresClasses.ts:80]:
```suggestion
```

**Review by 33Orange — CHANGES_REQUESTED** [review]
Almost done.

Lets improve readability

File: src/figuresClasses.ts [hunk 11]
```diff
@@ -1,19 +1,97 @@
...
+type Color = 'red' | 'green' | 'blue';
+
 export interface Figure {
+  shape: Shape;
+  color: Color;
+
+  a?: number;
+  b?: number;
+  c?: number;
+  radius?: number;
```
Comment (silchyn) [inline src/figuresClasses.ts:16]: does every figure have a, b, c, and radius?

File: src/figuresClasses.ts [hunk 12]
```diff
@@ -1,19 +1,97 @@
...
-export class Triangle {
+export class Triangle implements Figure {
+  public shape: Shape = Shape.triangle;
 
-}
+  constructor(
+    public color: Color,
+    public sideA: number,
+    public sideB: number,
+    public sideC: number,
```
Comment (silchyn) [inline src/figuresClasses.ts:28]:
```suggestion
    public a: number,
    public b: number,
    public c: number,
```

File: src/figuresClasses.ts [hunk 13]
```diff
@@ -1,19 +1,97 @@
...
 }
 
-export function getInfo(figure) {
+export class Rectangle implements Figure {
+  public shape: Shape = Shape.rectangle;
+
+  constructor(
+    public color: Color,
+    public sideA: number,
+    public sideB: number,
```
Comment (silchyn) [inline src/figuresClasses.ts:81]:
```suggestion
    public width: number,
    public height: number,
```

**Review by silchyn — CHANGES_REQUESTED** [review]

File: src/figuresClasses.ts [hunk 14]
```diff
@@ -1,19 +1,91 @@
...
+    const sumOfAllSides = a + b + c;
+    const sumOfSmallSides = sumOfAllSides - longestSide;
+
+    if (longestSide >= sumOfSmallSides) {
+      throw new Error('Triangle with given sides doesn\'t exist');
+    }
+
+    if (a <= 0 || b <= 0 || c <= 0) {
+      throw new Error('The triangle side must be greater than zero!');
+    }
```
Comment (silchyn) [inline src/figuresClasses.ts:34]: could be separate function

**Review by silchyn — APPROVED** [review]

**Pull request closed without merging**